### DISCUSSION TITLE
USB DFU Method

### DISCUSSION THREAD

**Discussion by Sathishkumar-rk** [zephyrproject-rtos/zephyr#45589] [Q&A]
Hi Team,

Is USB DFU method & linux dfu-util is supported for i.MXRT1060/1160

**Comment by danieldegrasse** [comment]
Hi @Sathishkumar-rk, USB DFU mode should be supported on the RT1060- see https://github.com/zephyrproject-rtos/zephyr/pull/45358 for the PR adding support for this feature. MCUBoot support on the iMX RT1160 is not upstream yet, but see this comment for the required changes to enable it: https://github.com/zephyrproject-rtos/zephyr/issues/45486#issuecomment-1125355982. With those updates, USB DFU mode works for me on the RT1160.

**Reply by Sathishkumar-rk** [reply]
Hi @danieldegrasse ,

**I was tried the build MCUBoot for iMXRT1060_EVK with USB DFU enabled state and seeing flash overflow error**

[285/295] Linking C executable zephyr\zephyr_pre0.elf
FAILED: zephyr/zephyr_pre0.elf zephyr/zephyr_pre0.map D:/Zephyrwork/zephyrproject/bootloader/mcuboot/build/zephyr/zephyr_pre0.map
cmd.exe /C "cd . && D:\Zephyrwork\zephyr-sdk-0.14.1\arm-zephyr-eabi\bin\arm-zephyr-eabi-gcc.exe   zephyr/CMakeFiles/zephyr_pre0.dir/misc/empty_file.c.obj -o zephyr\zephyr_pre0.elf  zephyr/CMakeFiles/offsets.dir/./arch/arm/core/offsets/offsets.c.obj  -Wl,-T  zephyr/linker_zephyr_pre0.cmd  -Wl,-Map=D:/Zephyrwork/zephyrproject/bootloader/mcuboot/build/zephyr/zephyr_pre0.map  -Wl,--whole-archive  app/libapp.a  zephyr/libzephyr.a  zephyr/arch/common/libarch__common.a  zephyr/arch/arch/arm/core/aarch32/libarch__arm__core__aarch32.a  zephyr/arch/arch/arm/core/aarch32/cortex_m/libarch__arm__core__aarch32__cortex_m.a  zephyr/arch/arch/arm/core/aarch32/mpu/libarch__arm__core__aarch32__mpu.a  zephyr/lib/libc/minimal/liblib__libc__minimal.a  zephyr/lib/posix/liblib__posix.a  zephyr/soc/arm/common/cortex_m/libsoc__arm__common__cortex_m.a  zephyr/subsys/dfu/boot/libsubsys__dfu__boot.a  zephyr/drivers/usb/device/libdrivers__usb__device.a  zephyr/drivers/clock_control/libdrivers__clock_control.a  zephyr/drivers/console/libdrivers__console.a  zephyr/drivers/gpio/libdrivers__gpio.a  zephyr/drivers/hwinfo/libdrivers__hwinfo.a  zephyr/drivers/memc/libdrivers__memc.a  zephyr/drivers/flash/libdrivers__flash.a  zephyr/drivers/serial/libdrivers__serial.a  zephyr/drivers/timer/libdrivers__timer.a  zephyr/drivers/pinctrl/libdrivers__pinctrl.a  modules/hal_nxp/hal_nxp/lib..__modules__hal__nxp.a  modules/mbedtls/libmodules__mbedtls.a  modules/mcuboot/libmcuboot_util.a  -Wl,--no-whole-archive  zephyr/kernel/libkernel.a  -L"d:/zephyrwork/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/thumb/v7e-m/nofp"  -LD:/Zephyrwork/zephyrproject/bootloader/mcuboot/build/zephyr  -lgcc  zephyr/libcode_relocation_source_lib.a  zephyr/arch/common/libisr_tables.a  -no-pie  -mcpu=cortex-m7  -mthumb  -mabi=aapcs  -mfp16-format=ieee  -Wl,--gc-sections  -Wl,--build-id=none  -Wl,--sort-common=descending  -Wl,--sort-section=alignment  -Wl,-u,_OffsetAbsSyms  -Wl,-u,_ConfigAbsSyms  -nostdlib  -static  -Wl,-X  -Wl,-N  -Wl,--orphan-handling=warn && cmd.exe /C "cd /D D:\Zephyrwork\zephyrproject\bootloader\mcuboot\build\zephyr && "C:\Program Files\CMake\bin\cmake.exe" -E echo ""
d:/zephyrwork/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.exe: zephyr\zephyr_pre0.elf **### section `sw_isr_table' will not fit in region `FLASH'**

d:/zephyrwork/zephyr-sdk-0.14.1/arm-zephyr-eabi/bin/../lib/gcc/arm-zephyr-eabi/10.3.0/../../../../arm-zephyr-eabi/bin/ld.exe: **### region `FLASH' overflowed by 15556 bytes**

collect2.exe: error: ld returned 1 exit status
ninja: build stopped: subcommand failed
![error](https://user-images.githubusercontent.com/105291276/169478962-b2131adc-e32d-4a8e-81c4-729358ce3737.png)
.

**Reply by danieldegrasse** [reply]
The default partition size for MCUBoot on the RT1060 EVK is 64KB, it is possible the mcuboot image you are building is exceeding that size. You can modify the partition layout to increase the size of the boot partition in the file `boards/arm/mimxrt1060_evk.dts`. Note that you will need to move each partition offset as well to ensure that the mcuboot image itself does not overlap with the bootloader.

For example, here is the diff for increasing the mcuboot partition to 256KB:

```diff
diff --git a/boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts b/boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts
index 70f5ffcebd..d311835e95 100644
--- a/boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts
+++ b/boards/arm/mimxrt1060_evk/mimxrt1060_evk.dts
@@ -126,23 +126,23 @@ arduino_serial: &lpuart3 {

                        boot_partition: partition@0 {
                                label = "mcuboot";
-                               reg = <0x00000000 DT_SIZE_K(64)>;
+                               reg = <0x00000000 DT_SIZE_K(256)>;
                        };
-                       slot0_partition: partition@10000 {
+                       slot0_partition: partition@40000 {
                                label = "image-0";
-                               reg = <0x00010000 DT_SIZE_M(3)>;
+                               reg = <0x00040000 DT_SIZE_M(3)>;
                        };
-                       slot1_partition: partition@310000 {
+                       slot1_partition: partition@340000 {
                                label = "image-1";
-                               reg = <0x00310000 DT_SIZE_M(3)>;
+                               reg = <0x00340000 DT_SIZE_M(3)>;
                        };
-                       scratch_partition: partition@610000 {
+                       scratch_partition: partition@640000 {
                                label = "image-scratch";
-                               reg = <0x00610000 DT_SIZE_K(128)>;
+                               reg = <0x00640000 DT_SIZE_K(128)>;
                        };
-                       storage_partition: partition@630000 {
+                       storage_partition: partition@660000 {
                                label = "storage";
-                               reg = <0x00630000 DT_SIZE_K(1856)>;
+                               reg = <0x00660000 DT_SIZE_K(1856)>;
                        };
                };
        };
```
